builder/amazon/common/step_run_spot_instance.go

```diff
@@ -139,101 +90,235 @@ func (s *StepRunSpotInstance) Run(ctx context.Context, state multistep.StateBag)
+
+	// Test if it is encoded already, and if not, encode it
+	if _, err := base64.StdEncoding.DecodeString(userData); err != nil {
+		log.Printf("[DEBUG] base64 encoding user data...")
+		userData = base64.StdEncoding.EncodeToString([]byte(userData))
 	}
+	return userData, nil
+}
+
+func (s *StepRunSpotInstance) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	ec2conn := state.Get("ec2").(*ec2.EC2)
+	ui := state.Get("ui").(packer.Ui)
+
+	ui.Say("Launching a spot AWS instance...")
 
-	runSpotResp, err := ec2conn.RequestSpotInstances(spotInstanceInput)
+	// Get and validate the source AMI
+	image, ok := state.Get("source_image").(*ec2.Image)
+	if !ok {
+		state.Put("error", fmt.Errorf("source_image type assertion failed"))
+		return multistep.ActionHalt
+	}
+	s.SourceAMI = *image.ImageId
+
+	if s.ExpectedRootDevice != "" && *image.RootDeviceType != s.ExpectedRootDevice {
+		state.Put("error", fmt.Errorf(
+			"The provided source AMI has an invalid root device type.\n"+
+				"Expected '%s', got '%s'.",
+			s.ExpectedRootDevice, *image.RootDeviceType))
+		return multistep.ActionHalt
+	}
+
+	azConfig := ""
+	if azRaw, ok := state.GetOk("availability_zone"); ok {
+		azConfig = azRaw.(string)
+	}
+	az := azConfig
+
+	ui.Message(fmt.Sprintf("Finding spot price for %s %s...",
+		s.SpotPriceProduct, s.InstanceType))
+	spotPrice, err := s.CalculateSpotPrice(az, ec2conn)
 	if err != nil {
-		err := fmt.Errorf("Error launching source spot instance: %s", err)
 		state.Put("error", err)
 		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
 
-	s.spotRequest = runSpotResp.SpotInstanceRequests[0]
+	ui.Message(fmt.Sprintf("Determined spot instance price of: %s.", spotPrice))
+
+	var instanceId string
+
+	ui.Say("Interpolating tags for spot instance...")
+	// s.Tags will tag the eventually launched instance
+	// s.SpotTags apply to the spot request itself, and do not automatically
+	// get applied to the spot instance that is launched once the request is
+	// fulfilled
+	if _, exists := s.Tags["Name"]; !exists {
+		s.Tags["Name"] = "Packer Builder"
+	}
 
-	spotRequestId := s.spotRequest.SpotInstanceRequestId
-	ui.Message(fmt.Sprintf("Waiting for spot request (%s) to become active...", *spotRequestId))
-	err = WaitUntilSpotRequestFulfilled(ctx, ec2conn, *spotRequestId)
+	// Convert tags from the tag map provided by the user into *ec2.Tag s
+	ec2Tags, err := s.Tags.EC2Tags(s.Ctx, *ec2conn.Config.Region, state)
 	if err != nil {
-		err := fmt.Errorf("Error waiting for spot request (%s) to become ready: %s", *spotRequestId, err)
+		err := fmt.Errorf("Error generating tags for source instance: %s", err)
 		state.Put("error", err)
 		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
+	// This prints the tags to the ui; it doesn't actually add them to the
+	// instance yet
+	ec2Tags.Report(ui)
+	// LaunchTemplateInstanceMarketOptionsRequest
+	// LaunchTemplateSpotMarketOptionsRequest
```

Comment:
Should we keep the 2 previous comments ?

Reply:
yep.